### DISCUSSION TITLE
Question on Memory Usage, maxmemory and cache_mode settings

### DISCUSSION THREAD

**Discussion by kaitalledo** [dragonflydb/dragonfly#1453] [General]
Hi Team,

We have successfully switched from Redis to DragonflyDB (1.4.0) on our production web application after testing it in our pre-production environment for sometime.  So far we are very happy with it but I just have one thing that we have noticed.  Previously, with Redis, memory usage was around 25GB (with maxmemory setting of 64GB and key eviction set to "volatile-lru") after we run a cache priming script. 

With dragonflydb, memory usage hovers around 1GB-1.5GB after priming.  Having said this, I'm not entirely sure if it's actually caching stuff or the cache_mode setting is doing something with it.  But as I understand it, based on the documentation, cache_mode should only kick in when the maxmemory is reached (also set to 64gb on dragonflydb).

Can someone help enlighten me?  Thank you in advance! Config of dragonflydb is below:

--pidfile=/var/run/dragonfly/dragonfly.pid
--log_dir=/var/log/dragonfly
--dir=/var/lib/dragonfly
--version_check=true
--bind=localhost
--maxmemory=64g
--cache_mode=true
--dbnum=16
--dbfilename=dump.rdb
--df_snapshot_format=false

**Comment by chakaz** [comment]
It's awesome that you've switched to Dragonfly, thanks for letting us know!
The fact that Dragonfly uses 25x less memory than Redis tells me that something is off. Do you use it in the exact same way, with roughly the same load? Indeed, `--cache_mode` should not have an effect when so much below `--maxmemory`.
Can you please share the output of `info`? (issue it inside `redis-cli`)

Also, a side note: I see that you're binding to `localhost`, does that mean you're running other services on the same machine? The reason I'm asking is because if that's the case, you may want to limit the number of threads used by Dragonfly, because by default it will use all, which could degrade other services using the same machine.

**Reply by kaitalledo** [reply]
Hello Chakaz,

Thanks for your response.  Yes we are using it exactly the same way we used Redis before. Below is the result of info:

127.0.0.1:6379> info
# Server
redis_version:6.2.11
dfly_version:df-v1.4.0
redis_mode:standalone
arch_bits:64
multiplexing_api:iouring
tcp_port:6379
uptime_in_seconds:14958
uptime_in_days:0

# Clients
connected_clients:1
client_read_buf_capacity:256
blocked_clients:0

# Memory
used_memory:77503600
used_memory_human:73.91MiB
used_memory_peak:77503600
comitted_memory:2372403200
used_memory_rss:1321345024
used_memory_rss_human:1.23GiB
object_used_memory:48618904
table_used_memory:26498976
num_buckets:631680
num_entries:49469
inline_keys:3002
strval_bytes:47192560
updateval_amount:0
listpack_blobs:0
listpack_bytes:0
small_string_bytes:24630680
pipeline_cache_bytes:0
maxmemory:68719476736
maxmemory_human:64.00GiB
cache_mode:cache

# Stats
total_connections_received:24043
total_commands_processed:953414
instantaneous_ops_per_sec:134
total_pipelined_commands:0
total_net_input_bytes:90049014
total_net_output_bytes:28928526
instantaneous_input_kbps:-1
instantaneous_output_kbps:-1
rejected_connections:-1
expired_keys:0
evicted_keys:0
hard_evictions:0
garbage_checked:0
garbage_collected:0
bump_ups:800386
stash_unloaded:0
oom_rejections:0
traverse_ttl_sec:0
delete_ttl_sec:0
keyspace_hits:800386
keyspace_misses:103551
total_reads_processed:945519
total_writes_processed:954961
async_writes_count:0
defrag_attempt_total:0
defrag_realloc_total:0
defrag_task_invocation_total:0

# Replication
role:master
connected_slaves:0
master_replid:a0fc7bbe305fa585b5be0aec0c894e58ba8d11af

# Keyspace
db0:keys=49468,expires=0,avg_ttl=-1
db4:keys=1,expires=0,avg_ttl=-1

# CPU
used_cpu_sys:475.112235
used_cpu_user:342.547171
used_cpu_sys_children:0.0
used_cpu_user_children:0.0
used_cpu_sys_main_thread:9.432727
used_cpu_user_main_thread:7.275755


As for binding it to localhost, that is correct.   We have put Dragonflydb in the same sever as where the webserver resides.  How can I limit the threads?

Cheers,

**Reply by chakaz** [reply]
I see that the DB has 49468 keys (+1 in db4), with 0 expired. Also, net received bytes is ~90mb, aligned with the current perceived consumption of 73mb. So it looks like this is just the data.. can you tell how many keys and other stats on the previous Redis deployment?

Re/ limiting threads - you could use `--proactor_threads=X` to limit Dragonfly to `X` threads

**Reply by kaitalledo** [reply]
Unfortunately, we have already decommissioned the Redis instance, so I am not able to get the details you are asing.

As for limiting the threads, I have now set it to 250 and the memory usage jumped to 6.5GB. 

Are there any other tips you can give me? :)

Cheers,

**Reply by chakaz** [reply]
I would actually suggest to lower that number:
* If you machine is dedicated to running Dragonfly, don't specify the number, which will have the effect of having 1 thread per hardware CPU. This is the ideal case.
* If you do run other services on that machine, calculate how many CPUs you need for them, and use the remainder for Dragonfly. For example, if your machine is running on a machine with 32 CPUs, and you need 10 for the other services, run dragonfly with `--proactor_threads=22` to let Dragonfly utilize only 22 CPUs

Running Dragonfly with more threads than physical CPUs will just reduce efficiency

**Reply by chakaz** [reply]
Oh, and the additional memory usage that you've witnessed is likely not due to more values stored in Dragonfly, but just the per-thread overhead x 250..

**Reply by kaitalledo** [reply]
All good now. Thank you so much!
